### PR TITLE
Add filtering

### PR DESCRIPTION
**Pull request by andyrichardson**
## About

Add filtering capability to timeline

## Todo
 - [ ] Add tests to TimelineRow
 - [ ] Add tests to Settings
 - [ ] Add fixture for Settings

### CODE REVIEW

File: src/panel/context/Timeline.tsx [hunk 1]
```diff
@@ -207,6 +210,10 @@ const useTimelineDomain = () => {
 export const TimelineProvider: FC = ({ children }) => {
   const { addMessageHandler } = useContext(DevtoolsContext);
   const domain = useTimelineDomain();
+  const [exchanges, setExchanges] = useState<string[]>([]);
+  const [filter, setFilter] = useState<{ source: string[] }>({
```
Comment (wgolledge) [inline src/panel/context/Timeline.tsx:214]: Nice! It should be pretty easy to add more types of filter

File: src/panel/pages/timeline/components/Settings.tsx [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  const [collapsed, setCollapsed] = useState(false);
+  const { exchanges, filter, setFilter } = useTimelineContext();
+
+  const handleExpandToggle = useCallback(() => setCollapsed((c) => !c), []);
+
+  const handleFilterToggle = useCallback(
+    (v: string) => () =>
+      setFilter((state) => ({
+        ...state,
+        source: state.source.includes(v)
```
Comment (wgolledge) [inline src/panel/pages/timeline/components/Settings.tsx:18]: Nit: could pull this out into a variable to make it a little more readable?

File: src/panel/pages/timeline/components/TimelineRow.tsx [hunk 3]
```diff
@@ -34,6 +29,7 @@ export const TimelineRow: FC<
               position: "absolute",
               left: scale(e.timestamp),
               transform: "translateX(-50%) translateY(-50%)",
+              visibility: filter.source.includes(e.source) || "hidden",
```
Comment (wgolledge) [inline src/panel/pages/timeline/components/TimelineRow.tsx:32]: If we just don't render the row in <Timeline /> does that cause issues when toggling on/off?

Comment (andyrichardson) [reply src/panel/pages/timeline/components/TimelineRow.tsx:32]: Initially I put the filtering logic in the context but then realised that filtering events from being shown on the timeline would need to be separate from filtering events that are in our state (for things such as network states).

The filter only filters the presentation of events on the timeline.

**Pull request merged**